content/6_Advanced/random.mdx

```diff
@@ -0,0 +1,318 @@
+- First we fix $S$ to be the number of "simulations" that we run.
+- Next, for each possible color $1 \leq c \leq n$, we define $v_{cj}$ as an independent random value within $[0,1]$, for each $1 \leq j \leq S$.
+- Then, we construct $S$ trees $T_i$, of the same structure as the initial tree, but with $v_{c_u i}$ marked on a node $u$.
+- For each query comparing $a,b,c,d$, we output $\texttt{Yes}$ if
+$$
+\sum_{i\leq S} \text{PathMin}(T_i,a,b) < \sum_{i\leq S} \text{PathMin}(T_i,c,d)
+$$
+and $\texttt{No}$ otherwise.
+
+This approach should "combine" the probabilities in a way that makes the more likely event (the path with more distinct values achieving a smaller value) exponentially more likely. For some intuition on this, if we flip a coin weighted with $\frac{2}{3}$ probability to be heads once, there is a $\frac{1}{3}$ probability it comes up tails more than heads. However, if we flip it $99$ times, the probability it comes up more tails more than heads becomes
+
+$$
+\sum_{k=50}^{99} \binom{99}{k} \left(\frac{1}{3}\right)^k \left(\frac{2}{3}\right)^{99-k} \approx .000309
+$$
+
+Calculating the exact probability that our approach works is more difficult than the coin toss example, but follows the same principle of failure becoming exponentially more unlikely across several trials. We can use some assumptions (notably the central limit theorem) to calculate a rough approximation of the probability of success. Here is a graph of the probability of passing the problem given a certain value of $S$. Note that the probability of success remains is largely the same across different counts of distinct colors along the path.
+
+![Graph](./assets/random-graph.png)
+
+The python code used to generate the estimates is shown below:
+
+```py
+import math
+from scipy.stats import norm
+
+
+def clt_probability(k, S):
+	"""
+	Computes the probability that the sum of S minima from 2k random draws
+	is less than the sum of S minima from k random draws, as approximated by the CLT.
+
+	The CLT approximation is given by:
+
+	   P(B<A) ≈ Φ(  (sqrt(S) * [k/((k+1)(2k+1)]) ) / sqrt( k/((k+1)^2(k+2)) + 2k/((2k+1)^2(2k+2)) ) )
+
+	:param k: The number of draws for each minimum:
+			- The first group uses k draws.
+	            	- The second group uses 2k draws.
+	:param S: The number of independent trials (sums over S minima).
+	:return: The approximated probability that the sum of the minima
+		 from 2k draws is less than that from k draws.
+	"""
+	numerator = math.sqrt(S) * k / ((k + 1) * (2 * k + 1))
+
+	var_X = k / (((k + 1) ** 2) * (k + 2))
+	var_Y = 2 * k / (((2 * k + 1) ** 2) * (2 * k + 2))
+
+	denominator = math.sqrt(var_X + var_Y)
+
+	z_score = numerator / denominator
+
+	probability = norm.cdf(z_score)
+
+	return probability
+```
+
+### Implementation
+
+By supporting path queries with HLD, we can solve this problem in $\mathcal{O}(S\cdot(n + m\log^2n))$.
+
+```cpp
+#include "bits/stdc++.h"
+using namespace std;
+
+using us = unsigned short;
+
+int read() {
```

Comment:
is this necessary?

Reply:
I also get TLE when submitting the in-module solution :/

https://qoj.ac/submission/971304

Reply:
what if it's just variation in judging times though?

Reply:
https://github.com/cpinitiative/usaco-guide/pull/5277#discussion_r2036495805

sure if it helps

Reply:
https://judge.yosupo.jp/submission/4542

might be more convenient to use

Reply:
how much time does read() save?